pedroslopez/whatsapp-web.js
# Emit chat state changed event when chat state changes (e.g. one contact starts typing)

# PR Details

The PR adds the functionality requested on #82 and on related issues: #1137, #2636, #3372.
This feature introduces the ability to detect chat states such as 'available', 'unavailable', 'typing' and 'recording', addressing a longstanding request from the community.

closes #82



## Description

3 main changes were made:
1. addition of the `CHAT_STATE_CHANGED` event on the Events object from `src/utiol/Contants.js`
2. addition of an exposed function to emit this particular event through the client object `src/Client.js:249`
3. addition of 2 functions at the window level
	- `injectPresenceModelInstance`: this function wraps the underlying modelClass from the Store.Presence object. the modelClass is used to generate all the models that represents a presence for a given chat.
	  This is done to "inject" the get and set properties on every single instance generate my the model class to be able to call the emit function every time the model chat state property is update 
	- `execPresenceModelInjection`: this function is called when the Store.Presence object becomes available (it takes a couple of milliseconds for it to load) and calls the `injectPresenceModelInstance` to inject every model already created.
	  The function also makes sure to maintain the prototype chain of every injected object to avoid breaking underlying functionality




## Motivation and Context

The main objective is to have one more way to improve user experience of apps built using whatapp-web.js. By providing a way to handle events related to chat state changes developers will be able to create interaction flows with a finer level of control.

An example use case would be: prevent an automatic response from being sent if the target user is typing, so it doesn't interrupt the user.



## To apply the PR run one of the following commands:
- **NPM**
```powershell
npm i github:lucas-almeida026/whatsapp-web.js#main
```
- **YARN**
```powershell
yarn add github:lucas-almeida026/whatsapp-web.js#main
```

## How Has This Been Tested

I tested manually using the `shell` command from the `package.json` file.

Actions:
- on the terminal: `npm run shell`
- after the shell loads: `client.on('chat_state_changed', console.log)`
- enter on my personal whatsapp account and open the chat with the connected account
- start typing
- observe the shell printing: `{ chatId: '*********@c.us', chatState: 'typing' }`
- stop typing on my phone
- observe the shell printing: `{ chatId: '*********@c.us', chatState: 'unavailable' }`


### Environment
- Machine OS: Linux - Ubuntu 24.04
- Phone OS: Android 11
- Library Version: 1.26.0
- WhatsApp Web Version: 2.3000.1018163899
- Puppeteer Version: ^18.2.1
- Browser Type and Version: Chromium 107.0.5296.0 - developer build 64-bit
- Node Version: 21.7.3
- NPM Version: 10.5.0

## Types of changes



- [ ] Dependency change

## Checklist



- [ ] ~~I have updated the documentation accordingly (index.d.ts).~~
- [x] I have updated the usage example accordingly (example.js)
